nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1051,29 +1063,102 @@ impl<'a> Planner<'a> {
+        let target_release = self.input.tuf_repo().description();
+        let mut out_of_date_zones = sleds
             .into_iter()
             .flat_map(|sled_id| {
-                let blueprint = &self.blueprint;
-                blueprint
+                let log = &self.log;
+                self.blueprint
                     .current_sled_zones(
                         sled_id,
                         BlueprintZoneDisposition::is_in_service,
                     )
                     .filter_map(move |zone| {
-                        (zone.image_source
-                            != blueprint
-                                .zone_image_source(zone.zone_type.kind()))
-                        .then(|| (sled_id, zone.clone()))
+                        let desired_image_source = match target_release
+                            .zone_image_source(zone.zone_type.kind())
+                        {
+                            Ok(source) => source,
+                            Err(err) => {
+                                // If we can't tell whether a zone is out of
+                                // date, assume it isn't.
+                                warn!(
+                                    log,
+                                    "cannot determine whether zone is \
+                                     out of date";
+                                    "zone" => ?zone,
+                                    InlineErrorChain::new(&err),
+                                );
+                                return None;
+                            }
+                        };
+                        if zone.image_source != desired_image_source {
+                            Some((sled_id, zone, desired_image_source))
+                        } else {
+                            None
+                        }
                     })
             })
-            .collect::<Vec<(SledUuid, BlueprintZoneConfig)>>();
-        if let Some((sled_id, zone)) = out_of_date_zones.first() {
-            return self.update_or_expunge_zone(*sled_id, zone);
+            .peekable();
+
+        // Before we filter out zones that can't be updated, do we have any out
+        // of date zones at all? We need this to explain why we didn't update
+        // any zones below, if we don't.
+        let have_out_of_date_zones = out_of_date_zones.peek().is_some();
+
+        // Of the out-of-date zones, filter out zones that can't be updated yet,
+        // either because they're not ready or because it wouldn't be safe to
+        // bounce them.
+        let mut updateable_zones =
+            out_of_date_zones.filter(|(_sled_id, zone, _new_image_source)| {
+                let kind = zone.zone_type.kind();
+                if !self.can_zone_be_shut_down_safely(kind) {
+                    return false;
+                }
+                match self.is_zone_ready_for_update(kind) {
```

Comment:
I think it's very likely we'll want to propagate "reasons" out from here, beyond just logging.

E.g., with cockroach, if we say "You cannot update because you have underreplicated ranges", that is a planner-decision that's probably worth documenting. Otherwise, the planner will just say "can't update now", which isn't that great - not even really identifying *which* zone needs work, or why.

TL;DR: We probably want a stronger return type than a boolean here, for both of these conditions? something like

```rust
enum ShutdownSafety {
  CanShutdown,
  CannotShutdown {
    reason: String,
  } 
}

enum UpdateReadiness {
  Ready,
  NotReady {
    reason: String,
  } 
}

```